Implement hill climbing

NicMcPhee opened the pull request:
This adds hill climbing, with examples.

There are several other minor refactorings and clean-ups along the way:
- We removed the return type from `RandomSearch` since it was just carrying an error, and upon examination it turned out that no error was ever generated.
- We added the `#[error(…)]` attribute to `HillClimberError` so it would implement `Display` and could be used with `anyhow`.

We decided not to implement parallelism here; see #3 for the related issue if you'd like to work on that later.

NicMcPhee committed:
Add `itertools` and `thiserror` as dependencies

NicMcPhee committed:
Get sequential hill-climbing working

NicMcPhee committed:
Finish up sequential hill-climber

This gets the sequential hill-climber working in a pretty reasonable way. Lots of help from esitsu@Twitch and @JustusFluegel on how to improve the structure of my initial implementation.

We spent a _lot_ of time flailing around attempts to use `max_by_key`, which just created all kinds of ownership and lifetime issues. Ultimately we just went with `max_by`, which works fine, but certainly isn't as nice as `max_by_key` would have been if we'd been able to get that to work.

Big thanks to esitsu@Twitch for the suggestion to use `slice::from_ref` to turn `&current_scored_best` into a `&[current_scored_best]` without copying. This really cleaned up the call to the `inspector` and allowed me to unwind some icky business that I had added to deal with references.

NicMcPhee committed:
Clean up a bunch of imports

NicMcPhee committed:
Remove stub for parallelism in hill climber

We decided to skip that for now since hill-climbing is a largely serial process.

NicMcPhee committed:
Removed return type from `RandomSearch`

Upon examination it became clear that `RandomSearch` never actually returned an error, so we removed the return type entirely since all it was doing was carrying the potential error.

NicMcPhee committed:
Add display/error to `HillClimberError`

I'd left out the `#[error(...)]` attribute on `HillClimberError` so it didn't work with `anyhow`. This fixes that.

NicMcPhee committed:
Implement hill-climbing search for an integer value

JustusFluegel commented on src/hill_climber.rs:
```diff
@@ -0,0 +1,108 @@
+// make the constraints more visible and (hopefully) make some of the error
+// messages more helpful for people new to Rust.
+where
+    Ge: Clone + std::fmt::Debug + Sync + Send,
```
Do we still use the debug bound?

NicMcPhee replied:
Nope we didn't. I'll remove that.

Actually, if we're not going to support parallelism here, we presumably don't need all this `Send + Sync` either. I'll remove that as well.

JustusFluegel replied:
I can't resolve this conversation (since I only have read access to this repo I think) so feel free to resolve it.

JustusFluegel reviewed (approved):
LGTM

NicMcPhee committed:
Remove unnecessary trait bounds

Thanks to @JustusFluegel for pointing that we didn't need the `Debug` bounds on `Ge` and `Sc` anymore. That triggered the observation that we also didn't need all the `Sync + Send` bounds on `HillClimber` since we're (currently) not implementing parallelism here.

JustusFluegel reviewed (approved):
LGTM
